metrics/heapster_test.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2016 Google Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package main
+
+import (
+	"crypto/tls"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net"
+	"net/http"
+	"testing"
+	"time"
+
+	"github.com/golang/glog"
+	"github.com/stretchr/testify/assert"
+
+	"k8s.io/heapster/metrics/apis/metrics/v1alpha1"
+	"k8s.io/heapster/metrics/cmd/heapster-apiserver/app"
+	"k8s.io/heapster/metrics/options"
+	"k8s.io/heapster/metrics/sinks/metric"
+	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/client/cache"
+)
+
+var metricsGroupVersion = v1alpha1.SchemeGroupVersion
+
+var metricsGroupVersionForDiscovery = unversioned.GroupVersionForDiscovery{
+	GroupVersion: metricsGroupVersion.String(),
+	Version:      metricsGroupVersion.Version,
+}
+
+var emptyMetricSink = &metricsink.MetricSink{}
+var emptyNodeLister = &cache.StoreToNodeLister{}
+var emptyPodLister = &cache.StoreToPodLister{}
+
+var testInsecurePort = 8080
+var testSecurePort = 6443
+
+func TestRunServer(t *testing.T) {
+	serverIP := fmt.Sprintf("http://localhost:%d", testInsecurePort)
+
+	go func() {
+		opt := getServerOptions()
+		opt.InsecurePort = testInsecurePort
+
+		server, err := app.NewHeapsterApiServer(opt, emptyMetricSink, emptyNodeLister, emptyPodLister)
+		if err != nil {
+			t.Fatalf("Could not create the API server: %v", err)
+		}
+		if err := server.RunServer(); err != nil {
+			t.Fatalf("Error in bringing up the server: %v", err)
+		}
+	}()
+	if err := waitForApiserverUp(serverIP); err != nil {
+		t.Fatalf("%v", err)
+	}
+	testSwaggerSpec(t, serverIP)
+	testAPIGroupList(t, serverIP)
+	testAPIGroup(t, serverIP)
+	testAPIResourceList(t, serverIP)
+}
+
+func TestRunSecureServer(t *testing.T) {
+	serverIP := fmt.Sprintf("https://localhost:%d", testSecurePort)
+
+	go func() {
+		opt := getServerOptions()
+		opt.InsecurePort = 0
+		opt.SecurePort = testSecurePort
+		opt.CertDirectory = "./tmp"
```

Comment:
The default value here is "/var/run/kubernetes", which fails with permission denied, so I'm not sure what should be used here. Is there any particular place the certificates should go to?